### DISCUSSION TITLE
How to manage external data efficiently?

### DISCUSSION THREAD

**Discussion by piseabhijeet** [treeverse/dvc#5154] [Help] [answered]
Hi,
I have configured dvc for data version control in my Git project. I have some data on a Temp folder **D:\Temp\Data**.

I add it as an external dependency using the command:
**dvc add --external D:\Temp\Data**
which produces a **Data.dvc** file 

I now define a dvc run which copies the data from D:\Temp\Data and keeps it in my Workspace. It produces a .yaml and .lock file.

I now change the original data on D:\Temp\Data by adding new files to it. 

When i do a dvc status, it shows files have been changed, but i dont know how to add those the version control. I tried dvc import but it works on only DVC/Git based repos not on data on local storage or S3.

I do not want to push the data because it bloats the storage but i want to track the same using external dependency only.

Any help would be appreciated :)

**Comment by pmrowla** [comment, marked as answer]
To add the new version to DVC and update `Data.dvc` you would just rerun
```
dvc add --external D:\Temp\Data
```
(This is the same way that you update regular local workspace data which has been tracked via `dvc add`).

Using external data is a fairly advanced DVC feature and generally not recommended, can you share a bit more about your specific needs/use case here?

**Reply by piseabhijeet** [reply]
Yes i tried the same command, but it says Data.dvc is already defined.

The scenario:
I have a pool of data on S3 (s3://bucket/captured_data) which gets updated every couple of days (ex. day 1 it has 100 images, day 3 it will have 200 images, day 5 it will have 300 images and so on), i need to version the folder 'captured_data'. On day 1 if i run:
**dvc add --external s3:/bucket/captured_data ** command and use a dvc run pipeline, it will create a captured_data.dvc file, however, on day3, the orignal data source is added with more data, how do i version that? 

Using a remote storage option is inelegant as the data is already on remote storage, i will be duplicating the data by doing **dvc push**) 

I did try the external dependency as well but the cache on S3 is also flooded with duplicate data.

Is there a better way of doing this?

**Reply by piseabhijeet** [reply]
@pmrowla 

Also, when you say "external data is a fairly advanced DVC feature and generally not recommended" what is the use case for the external data then? 

My data lies on S3 and i train my models using an AWS EC2 instance, I do not want to pull data from S3, version it and push it to S3 again using 'dvc push' and remote storage option. Whereas in 'external dependency' scenario, i point the data on S3, configure a cache on a different folder in S3 (which also has copies of data in a dvc specific format). Now when new images are added to the same folder, how is cache updated?

**Reply by pmrowla** [reply]
In the external output/dependency scenario, your external data location on S3 is treated as something like an extension of your local workspace, except on S3. Different versions of files tracked by DVC still need to be cached, and when you do `dvc add --external s3://...`, that external file will be copied into your S3 cache location, the same thing that happens when you track local files using `dvc add`. The difference is that for external data it would use your S3 cache location rather than the local `.dvc/cache`.

**Reply by pmrowla** [reply]
Regarding use cases, it's really more of a workaround for when the workspace itself has to be external, such as when data can't leave a specific S3 bucket due to privacy/security/etc reasons.

**Reply by piseabhijeet** [reply]
Understood. Thank you for the explanation.

So if the data gets updated on **s3://.../data** , dvc status shows that **data** is changed, but the **data.dvc** is not updated, when i run the same **dvc add --external s3://.../data** it says the data.dvc file is already present. 

Also the **s3://.../data** can contain folders such as **v1, v2, v3 ...** how to version those as and when the folders are added so that when i do a dvc run, all the folders are retrieved ?

**Reply by pmrowla** [reply]
If you `dvc add` a directory, all of the files and subfolders in that directory will be included (unless they are listed in `.dvcignore`).

Can you post the full error from when you run `dvc add --external ...`? Also, how are you referring to `data` in your run pipeline (maybe just post your `dvc.yaml`)?

**Reply by piseabhijeet** [reply]
dvc.yaml:

stages:
  download_photos_2_for_training:
    cmd: aws s3 cp s3://bucket/photos2/ .\Training_Data --recursive
    deps:
    - s3://bucket/photos_2/
    outs:
    - .\Training_Data


Now i upload new data in photos_2 folder,

The log after i do **dvc status**:

download_photos_2_for_training:
        changed deps:
                modified:           s3://bucket/photos_2/
photos_2.dvc:
        changed outs:
                modified:           s3://bucket/photos_2

after this i run **dvc add --external s3://bucket/photos_2**
the output is:
100% Add|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|1/1 [02:34, 154.20s/file]

**To track the changes with git, run:

        git add photos_2.dvc**

new files get added to cache folder, ideally only the files which are added should be reflected, but i see new files as well, did it duplicate the earlier data again?

Now when i add that and again update the stage:
**dvc run -n download_photos_2_for_training -d s3://bucket/photos_2/ -o .\Training_Data aws s3 cp s3://bucket/photos_2/ .\Training_Data --recursive**

it shows:
**ERROR: Stage 'download_photos_2_for_training' already exists in 'dvc.yaml'. Use '--force' to overwrite.**

is it okay to overwrite the stage? Can i skip dvc.yaml run stage and handle the data pull using aws cp inside the data getter code?

**Reply by pmrowla** [reply]
`dvc run` is for creating (or modifying) stages in `dvc.yaml`. If you have already created your stage, you should be using [dvc repro](https://dvc.org/doc/command-reference/repro#repro) to reproduce the stages defined in `dvc.yaml`.

**Reply by piseabhijeet** [reply]
No I mean, since i am using the same location to copy data from i need not create additional stages for new data uploaded eg. v2, v3, v4 under the same folder right? I can use the same dvc.yaml file and just change the command, eg for first commit, i will use **aws cp s3://bucket/photos_2/v1 ./Training_Data**, for the second commit, using the same stage and the same yaml file but with a different cp command **aws cp s3://bucket/photos_2/v1 ./Training_Data/v1 aws cp s3://bucket/photos_2/v2 ./Training_Data/v2** does this work?

**Reply by pmrowla** [reply]
You don't need to create new stages. You should use `dvc run` once to define a stage that runs the command `aws s3 cp s3://bucket/photos_2/ .\Training_Data --recursive`. After you have done that, whenever your input data has changed, you just run `dvc repro`.

**Reply by pmrowla** [reply]
Also, I still don't really understand why you need to use external here.

When you have:
```
outs:
- .\Training_Data
```
It will tell DVC to track the local workspace directory `Training_Data/` as an output. This means that every time you are pulling data from S3 via the external dependency, it is still going to end up being versioned and cached as a regular local DVC output. So you will have duplicated data in both the S3 cache and the local cache. If you don't want DVC to track/cache the local data, you should either not mark it as an output at all, or explicitly tell DVC that it is an uncached output (refer to the docs for `dvc run -O/--outs-no-cache`).

**Reply by piseabhijeet** [reply]
Oh okay, got it, thanks @pmrowla. I thought we need to create an output stage after an external data is specified. Then in that case there is no need to redefine pipelines again. Is there a good reference on how to use an external source efficiently?

**Reply by pmrowla** [reply]
There is really only the official documentation. Our support for external data is likely going to change, because a vast majority of the time what people are really looking for is #4520, which has already been mentioned. And in fact, I'm still fairly sure that feature is actually what you want here (it does what your original pipeline was trying to do).

**Reply by piseabhijeet** [reply]
@pmrowla 
Thank you for the prompt response. Is there a timeline as to when this feature will be available?

Other thing, not sure if it is relevant, is there a support for dvc with mlflow? By support i mean currently dvc has data processing as well as experiment reproduction stages and mlflow also has the same, are there any best practices of using dvc as data version tool + mlflow as MLOps tool?

**Reply by pmrowla** [reply]
It's fairly high on our priorities list given that it is a frequently requested feature, but there is currently no estimate on when it will be available.

Regarding MLFlow, there are a few projects that use both DVC and MLFlow, one example is https://medium.com/ixorthink/our-machine-learning-workflow-dvc-mlflow-and-training-in-docker-containers-5b9c80cdf804, you'll have to look around for some others.

**Reply by piseabhijeet** [reply]
Than you @pmrowla , the discussion helped me to understand things in a better manner :)

**Comment by efiop** [comment]
@piseabhijeet Looks like you are looking for https://github.com/iterative/dvc/issues/4520 , which is on our todo list right now. The current workflow to achieve that is to just copy the file you wish to add into your workspace and then `dvc add data` there.

**Reply by piseabhijeet** [reply]
@efiop Thank you for the update, My scenario is explained in 
https://github.com/iterative/dvc/discussions/5154#discussioncomment-237728

I need to track data on S3 without duplicating it on S3 using dvc push (remote storage scenario) or on cache (external dependency scenario) and in the same .dvc workspace file
